README.md

```diff
@@ -1,142 +1,333 @@
-providing a great flexibility for rich and interactive UI, this may introduce overhead for highly
-dynamic content. So, the implementation try to minimize the number of ansi escapes sequences
-generated to draw the updated UI. In practice, given the speed of `Rust` the overhead rather comes
-from the terminal emulator than the library itself.
-
-Moreover, the library does not provide any input handling nor any event system and you may rely on
-the previously cited libraries to achieve such features.
-
-We keep a [CHANGELOG](./CHANGELOG.md) generated by [git-cliff](https://github.com/orhun/git-cliff)
-utilizing [Conventional Commits](https://www.conventionalcommits.org/).
+- [Ratatui Book] - explains the library's concepts and provides step-by-step tutorials
+- [Examples] - a collection of examples that demonstrate how to use the library.
+- [API Documentation] - the full API documentation for the library on docs.rs.
+- [Changelog] - generated by [git-cliff](https://github.com/orhun/git-cliff) utilizing [Conventional
+  Commits](https://www.conventionalcommits.org/).
+- [Contributing] - Please read this if you are interested in contributing to the project.
 
 ## Quickstart
 
 The following example demonstrates the minimal amount of code necessary to setup a terminal and
 render "Hello World!". The full code for this example which contains a little more detail is in
-[hello_world.rs](./examples/hello_world.rs). For more guidance on how to create Ratatui apps, see
-the [Docs](https://docs.rs/ratatui) and [Examples](#examples). There is also a starter template
-available at [rust-tui-template](https://github.com/ratatui-org/rust-tui-template).
+[hello_world.rs]. For more guidance on different ways to structure your application see the
+[Application Patterns](https://ratatui.rs/concepts/application-patterns/index.html) and [Hello
+World tutorial](https://ratatui-org.github.io/ratatui-book/tutorial/hello-world/index.html)
+sections in the [Ratatui Book] and the various [Examples]. There are also several starter
+templates available:
+
+- [rust-tui-template](https://github.com/ratatui-org/rust-tui-template).
+- [ratatui-async-template](https://ratatui-org.github.io/ratatui-async-template/) (book and
+  template)
+- [simple-tui-rs](https://github.com/pmsanford/simple-tui-rs)
+
+Every application built with `ratatui` needs to implement the following steps:
+
+- Initialize the terminal
+- A main loop to:
+  - Handle input events
+  - Draw the UI
+- Restore the terminal state
 
 ```rust
-fn main() -> Result<(), Box<dyn Error>> {
-    let mut terminal = setup_terminal()?;
-    run(&mut terminal)?;
-    restore_terminal(&mut terminal)?;
+use std::io;
+
+fn main() -> io::Result<()> {
+    let mut terminal = init_terminal()?;
+    let mut should_quit = false;
+    while !should_quit {
+        terminal.draw(ui)?;
+        should_quit = handle_events()?;
+    }
+    restore_terminal()?;
```

Comment:
This feels like too much information for the landing page (or too little depending on how you look at it). 

I'd prefer to move all this content in the `Quickstart` section here into a `quickstart.md` section in the book, and link to it from this landing page. And either have no code in the landing page or the smallest "hello world" example.

In my opinion, quickstart sections are for "pattern recognition", i.e. what is the kind of code I'll have to write to use this crate. 

I'm curious to hear what others think too.

Reply:
> This feels like too much information for the landing page (or too little depending on how you look at it).

I strongly disagree. When I go to a docs.rs page I want to understand how the major modules, types, traits and methods of a crate all fit together, and then jump off from there into the details. For reference level material to be effective, it needs to be well grounded with examples that help show why you're jumping to the other parts of the documentation. The least effective crate docs I've read have been those that don't take this step.

> I'd prefer to move all this content in the Quickstart section here into a quickstart.md section in the book, and link to it from this landing page. And either have no code in the landing page or the smallest "hello world" example.

I see the Ratatui book very much as a secondary source. Even when it has the best quality information I should be able to answer any reasonable question about how to use Ratatui using only the docs.rs for the version of the crate. Hard disagree on no code. And this is the smallest hello world that anyone should write (ignoring the layout and style sections)

> In my opinion, quickstart sections are for "pattern recognition", i.e. what is the kind of code I'll have to write to use this crate.

For ratatui, those critical things to understand for any app are initialization / restore, main loop, input handling, drawing the UI. Styling and layout round out the second tier of things which are perhaps not critical, but still necessary. Having a section and example method for each of those areas, with links to the rest of the docs and to the book seems "just right" to me (I'd neither want to put it all in a single file, or break it down any further)

I see tutorial sections in the book as teaching someone how to create an application, not for finding out the available modules and types that I need to use in Ratatui app. Because of this, they are low-signal high-noise pointers to reference docs. I have to read a lot of information before I find the information I need. This overview page should be the opposite of that (high-signal, low noise)

I chose the density of information pretty carefully in writing up the quickstart example here. Right now the existing docs.rs page:
- has too much detail on how to use alternate backends. replaced with a link
- Shows cargo.toml in addition to cargo add - unnecessary to show two ways of doing this
- Shows how to create a backend for multiple backends
- merges information about raw mode, alternate screen, and rendering the UI into one code block that is longer than a page. Shows some pretty questionable app choices
- jumps to layout in a way that doesn't link with the previous example
- doesn't cover the prelude, styling, style shortcuts (all of these are useful to call out as early as possible in the docs to build awareness)

My goals with the replacement doc was:
- avoid describing too much detail about why things work, preferring to show the options and jump to the pages that go into detail (for backends, etc.)
- Keep each example and section concise and focused rather than page long walls of text
- Label each section to make the docs easy to read for non-linear reading
- Ensure that we show `key_event.kind == KeyEventKind::KeyPress` to help avoid that question
- Show the prelude and style shortcuts to make user's code shorter and easier to read

I also really don't like jumping back and forth between two websites to get the information that I need to use a library. This is especially a problem as the life of the book website compared to the life of a docs.rs page is different. The docs for 0.24.0 will be the docs for 0.24.0 forever, the Ratatui book in a couple of years will probably be quite different.

In summary, I think this is probably goldilocks when it comes to docs, rather than too much / too little.

Reply:
I think this PR is an improvement over the current state, which is great! But I'm just not convinced it is really necessary. 

> I also really don't like jumping back and forth between two websites to get the information that I need to use a library. 

I agree with this a lot and I think this is my main reasoning for having everything in the website/book. There's ways to point to different versions of the docs (e.g. different folders for different versions) if we wanted to do that.

I think of the docs.rs landing page similar to a README.md file, and I think such a page should just be the one stop shop to find all the links to useful information, not necessary provide the actual useful information. I also think this comes down to personal preference. 

If we are optimizing for new users, I don't like the idea of having code blocks that use functions that are used later. If I were reading this for the first time, I bet I'd be confused. Is `init_terminal()` a function provided by the `ratatui` crate? Why is it not imported? If I were new to rust, there'd be even more things that would confuse me: what's `ui<B: Backend>()`? Do I have to write scary code like that?

I also agree that we should avoid going into too much detail which is why I think the _smallest_ example possible would be best. This might just be me but most of the time when I'm scanning through the docs.rs landing page for a crate, I'm skimming through all the code in the page to get a sense for what the crate is about. All text surrounding the code is secondary to me when I'm just exploring. If I'm actually evaluating using it, I start looking through the detailed docs and the examples. The landing page is not useful in that regard.

See for example [`serde`'s README.md](https://github.com/serde-rs/serde/blob/ddc1ee564b33aa584e5a66817aafb27c3265b212/README.md) as an example of a short and sweet README.
It has a fully working example in 20 or so lines of code, including comments that show that you'd expect to get if you ran the code. 

Other landing pages also show a simple program from start to finish, and show the output when you run that program. [clap](https://docs.rs/clap/4.4.4/clap/) for example shows what happens when you run `demo --help`. I'd argue that with the current PR's quickstart section, it is not easy to predict what the output would be. It would benefit from a screenshot or the output copy pasted into a codeblock.

All that said, this is not something I feel too strongly about and I do like the content quickstart section from the PR :) I'm just not convinced it is doing the most useful work as part of the landing page on docs.rs. But in case someone doesn't care, they'll just scroll past it all, so it's fine with having it if you think it is important. I'm sure people read docs differently and we'd want to cater to all needs, and my perspective here might be biased :)